Add support for `$expr` operator

divyajnu08 opened the pull request:
<!--

Thank you for proposing a pull request. This template will guide you through the essential steps necessary for a pull request.
Make sure that:

-->

- [ ] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [ ] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [ ] You submit test cases (unit or integration tests) that back your changes.
- [ ] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

mp911de commented:
Thanks for submitting a pull request. We have dedicated operator classes (e.g. `ArithmeticOperators`) to build `AggregationExpression` which can be used with `$project`. For example:

```java
Aggregation.addFields().addField("foo")
		.withValue(ArithmeticOperators.valueOf("otherField").stdDevPop()).build();

Aggregation.project().and(ArithmeticOperators.valueOf("otherField").stdDevPop()).as("otherFieldStdDevPop"));
```

Accepting `AggregationExpression` and field references in stages that do not accept these yet makes still sense. I suggest removing `stdDevPop` and `stdDevSamp` operators from this pull request as we have support for these already and updating this PR to accept `AggregationExpression` in stages that do not accept these yet.

divyajnu08 commented:
Hi Mark

I have removed the operators and added unit tests using ArithmeticOperators.
I think all operators support this operator.
Also kindly check the newly added $expr operator and changes in match operator if its correct.

Thanks

mp911de commented:
Thank you for your contribution. That's merged and polished now. We need to follow up with #3587 to enable a better usage of `$expr` with our Query infrastructure.
